src/protocolsupport/zplatform/pe/PECraftingManager.java

```diff
@@ -0,0 +1,145 @@
+                ItemStackWrapper[] required = new ItemStackWrapper[width * height];
+                for (int z = 0; z < height; ++z) {
+                    for (int x = 0; x < width; ++x) {
+                        int i = z * x;
+                        char key = pattern[z].charAt(x);
+                        try {
+                            org.bukkit.inventory.ItemStack stack = map.get(key);
+                            required[i] = stack == null || stack.getTypeId() < 1 ? ServerPlatform.get().getWrapperFactory().createItemStack(0) : fromBukkitStack(stack);
+                        } catch (NullPointerException e)    {
+                            ProtocolSupport.logInfo("[WARN] Unable to locate key " + key + " for recipe with output " + recipe.getResult().toString());
+                            return;
+                        }
+                    }
+                }
+                addRecipeShaped(fromBukkitStack(shapedRecipe.getResult()), width, height, required);
+            } else if (recipe instanceof ShapelessRecipe)   {
+                ShapelessRecipe shapelessRecipe = (ShapelessRecipe) recipe;
+                ItemStackWrapper[] required = new ItemStackWrapper[shapelessRecipe.getIngredientList().size()];
+                for (int i = 0; i < required.length; i++)   {
+                    required[i] = fromBukkitStack(shapelessRecipe.getIngredientList().get(i));
+                }
+                addRecipeShapeless(fromBukkitStack(recipe.getResult()), required);
+            } else if (recipe instanceof FurnaceRecipe) {
+                FurnaceRecipe shapelessRecipe = (FurnaceRecipe) recipe;
+                addRecipeFurnace(fromBukkitStack(shapelessRecipe.getResult()), fromBukkitStack(shapelessRecipe.getInput()));
+            } else {
+                ProtocolSupport.logInfo("unknown recipe type: " + recipe.getClass().getCanonicalName());
+            }
+        });
+
+        byte[] cached = byteBuf.array();
+        byteBuf.clear();
+
+        VarNumberSerializer.writeVarInt(byteBuf, recipeCount);
+        byteBuf.writeBytes(cached);
+
+        ProtocolSupport.logInfo("Done! Processed " + recipeCount + " recipes!");
+    }
+
+    public ItemStackWrapper fromBukkitStack(org.bukkit.inventory.ItemStack stack) {
+        return ServerPlatform.get().getWrapperFactory().createItemStack(stack.getType());
+    }
+
+    public void addRecipeShaped(ItemStackWrapper output, int width, int height, ItemStackWrapper[] required)    {
+        recipeCount++;
+        VarNumberSerializer.writeSVarInt(byteBuf, 1); //type
+        VarNumberSerializer.writeSVarInt(byteBuf, width);
+        VarNumberSerializer.writeSVarInt(byteBuf, height);
+        for (ItemStackWrapper stack : required) {
+            ItemStackSerializer.writeItemStack(byteBuf, ProtocolVersion.MINECRAFT_PE, "en_US", stack, true);
+        }
+        VarNumberSerializer.writeVarInt(byteBuf, 1); //not sure but pocketmine has it
```

Comment:
Need to find out what it is then

Reply:
according to dktapps it's the recipe result count

Reply:
Then this infomation should be extracted from result itemstack i think

Reply:
nope, it's in case you have recipes that have multiple outputs

Reply:
Change the comment then, so we don't have to dig up that info when updating